command/agent/agent.go

```diff
@@ -165,7 +165,7 @@ func (c *cmd) run(args []string) int {
 		return 1
 	}
 
-	bd, err := agent.NewBaseDeps(loader, logGate, nil)
+	bd, _, err := agent.NewBaseDeps(loader, logGate, nil)
```

Comment:
Should we store the `cleanup` func and execute it on agent shutdown? Then it wouldn't be a special case for tests? I realise our general assumption is that agents outside of tests always live for the whole process lifetime but still I feel like we should aim to move towards a world where agents don't just leak by default and need special cases to clean up in tests!

If we did that both here and in `OverrideDeps` which is called by `TestAgent` then would be a small step towards agents actually cleaning themselves up properly on shutdown all the time rather than needing testing special cases?

No strong opinon though, there may be other downsided to that approach I'm not thinking of.

Reply:
Agreed! I almost added `defer cleanup()` here, but wasn't sure if there was much point as the process exits once the agent has shut down anyway.

I like the idea in your other comment of adding a `Close` method to `BaseDeps`, which we could call from `ShutdownAgent` to solve both cases. We could even move the following 2 lines there too:

https://github.com/hashicorp/consul/blob/9108aace2e05f16179ea4d6471fdc8dd067c67bc/agent/agent.go#L1601-L1604